File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.java
/*************************************************************************/
/*  GodotEditor.java                                                     */
/*************************************************************************/
/*                       This file is part of:                           */
/*                           GODOT ENGINE                                */
/*                      https://godotengine.org                          */
/*************************************************************************/
/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
/*                                                                       */
/* Permission is hereby granted, free of charge, to any person obtaining */
/* a copy of this software and associated documentation files (the       */
/* "Software"), to deal in the Software without restriction, including   */
/* without limitation the rights to use, copy, modify, merge, publish,   */
/* distribute, sublicense, and/or sell copies of the Software, and to    */
/* permit persons to whom the Software is furnished to do so, subject to */
/* the following conditions:                                             */
/*                                                                       */
/* The above copyright notice and this permission notice shall be        */
/* included in all copies or substantial portions of the Software.       */
/*                                                                       */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
/*************************************************************************/

package org.godotengine.editor;

import org.godotengine.godot.FullScreenGodotApp;
import org.godotengine.godot.utils.PermissionsUtil;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Debug;

import androidx.annotation.Nullable;
import androidx.window.layout.WindowMetrics;
import androidx.window.layout.WindowMetricsCalculator;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Base class for the Godot Android Editor activities.
 *
 * This provides the basic templates for the activities making up this application.
 * Each derived activity runs in its own process, which enable up to have several instances of
 * the Godot engine up and running at the same time.
 *
 * It also plays the role of the primary editor window.
 */
public class GodotEditor extends FullScreenGodotApp {
	private static final boolean WAIT_FOR_DEBUGGER = false;
	private static final String COMMAND_LINE_PARAMS = "command_line_params";

	private static final String EDITOR_ARG = "--editor";
	private static final String PROJECT_MANAGER_ARG = "--project-manager";

	private final List<String> commandLineParams = new ArrayList<>();

	@Override
	public void onCreate(Bundle savedInstanceState) {
		PermissionsUtil.requestManifestPermissions(this);

		String[] params = getIntent().getStringArrayExtra(COMMAND_LINE_PARAMS);
		updateCommandLineParams(params);

		if (BuildConfig.BUILD_TYPE.equals("debug") && WAIT_FOR_DEBUGGER) {
			Debug.waitForDebugger();
		}
		super.onCreate(savedInstanceState);
	}

	private void updateCommandLineParams(@Nullable String[] args) {
		// Update the list of command line params with the new args
		commandLineParams.clear();
		if (args != null && args.length > 0) {
			commandLineParams.addAll(Arrays.asList(args));
		}
	}

	@Override
	public List<String> getCommandLine() {
		return commandLineParams;
	}

	@Override
	public void onNewGodotInstanceRequested(String[] args) {
		// Parse the arguments to figure out which activity to start.
		Class<?> targetClass = GodotGame.class;
		// Whether we should launch the new godot instance in an adjacent window
		// https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_LAUNCH_ADJACENT
		boolean launchAdjacent = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && (isInMultiWindowMode() || isLargeScreen());

		for (String arg : args) {
			if (EDITOR_ARG.equals(arg)) {
				targetClass = GodotEditor.class;
				launchAdjacent = false;
				break;
			}

			if (PROJECT_MANAGER_ARG.equals(arg)) {
				targetClass = GodotProjectManager.class;
				launchAdjacent = false;
				break;
			}
		}

		// Launch a new activity
		Intent newInstance = new Intent(this, targetClass)
									 .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
									 .putExtra(COMMAND_LINE_PARAMS, args);
		if (launchAdjacent) {
			newInstance.addFlags(Intent.FLAG_ACTIVITY_LAUNCH_ADJACENT);
		}
		startActivity(newInstance);
	}

	protected boolean isLargeScreen() {
		WindowMetrics metrics =
				WindowMetricsCalculator.getOrCreate().computeMaximumWindowMetrics(this);

		// Get the screen's density scale
		float scale = getResources().getDisplayMetrics().density;

		// Get the minimum window size
		float minSize = Math.min(metrics.getBounds().width(), metrics.getBounds().height());
		float minSizeDp = minSize / scale;
		return minSizeDp >= 840f; // Correspond to the EXPANDED window size class.
	}

	@Override
	public void setRequestedOrientation(int requestedOrientation) {
		if (!overrideOrientationRequest()) {
			super.setRequestedOrientation(requestedOrientation);
		}
	}

	/**
	 * The Godot Android Editor sets its own orientation via its AndroidManifest
	 */
	protected boolean overrideOrientationRequest() {
		return true;
	}
}

File: platform/android/java/editor/src/horizonos/java/org/godotengine/editor/GodotEditor.kt
/**************************************************************************/
/*  GodotEditor.kt                                                        */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

package org.godotengine.editor

import org.godotengine.godot.GodotLib
import org.godotengine.godot.utils.isNativeXRDevice

/**
 * Primary window of the Godot Editor.
 *
 * This is the implementation of the editor used when running on HorizonOS devices.
 */
open class GodotEditor : BaseGodotEditor() {

	companion object {
		private val TAG = GodotEditor::class.java.simpleName

		internal val XR_RUN_GAME_INFO = EditorWindowInfo(GodotXRGame::class.java, 1667, ":GodotXRGame")

		internal val USE_SCENE_PERMISSIONS = listOf("com.oculus.permission.USE_SCENE", "horizonos.permission.USE_SCENE")
	}

	override fun getExcludedPermissions(): MutableSet<String> {
		val excludedPermissions = super.getExcludedPermissions()
		// The USE_SCENE permission is requested when the "xr/openxr/enabled" project setting
		// is enabled.
		excludedPermissions.addAll(USE_SCENE_PERMISSIONS)
		return excludedPermissions
	}

	override fun retrieveEditorWindowInfo(args: Array<String>): EditorWindowInfo {
		var hasEditor = false
		var xrModeOn = false

		var i = 0
		while (i < args.size) {
			when (args[i++]) {
				EDITOR_ARG, EDITOR_ARG_SHORT, EDITOR_PROJECT_MANAGER_ARG, EDITOR_PROJECT_MANAGER_ARG_SHORT -> hasEditor = true
				XR_MODE_ARG -> {
					val argValue = args[i++]
					xrModeOn = xrModeOn || ("on" == argValue)
				}
			}
		}

		return if (hasEditor) {
			EDITOR_MAIN_INFO
		} else {
			val openxrEnabled = GodotLib.getGlobal("xr/openxr/enabled").toBoolean()
			if (openxrEnabled && isNativeXRDevice()) {
				XR_RUN_GAME_INFO
			} else {
				RUN_GAME_INFO
			}
		}
	}

	override fun getEditorWindowInfoForInstanceId(instanceId: Int): EditorWindowInfo? {
		return when (instanceId) {
			XR_RUN_GAME_INFO.windowId -> XR_RUN_GAME_INFO
			else -> super.getEditorWindowInfoForInstanceId(instanceId)
		}
	}
}
